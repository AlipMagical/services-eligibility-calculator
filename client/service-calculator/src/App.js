/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import './i18n';

import React, { createContext, useState } from 'react';

import { ANALYTICS_ID } from './common/config';
import { Body } from './components/Body';
import { DataContext } from './DataContext';
import Footer from './components/Footer';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import { Router } from 'react-router-dom';
import ScrollToTop from './common/utils/ScrollToTop';
import { createBrowserHistory } from 'history';

export const FilterContext = createContext();

export const createInitialFilterState = () => ({
  income: undefined,
  interests: {},
  people: {}
});

const history = createBrowserHistory();

export default function App(props) {
  const [analyticsInitialized, setAnalyticsInitialized] = useState(false);
  const [lastSearch, setLastSearch] = useState('');
  const [filterState, setFilterState] = useState(createInitialFilterState());
  const resetFilterState = () => setFilterState(createInitialFilterState());

  if (!analyticsInitialized) {
    ReactGA.initialize(ANALYTICS_ID, { testMode: props.testMode });
    setAnalyticsInitialized(true);
  }

  return (
    <DataContext.StateProvider>
      <FilterContext.Provider
        value={{
          filterState,
          setFilterState,
          resetFilterState,
          lastSearch,
          setLastSearch
        }}
      >
        <Router history={history}>
          <ScrollToTop />
          <Body />
          <Footer />
        </Router>
      </FilterContext.Provider>
    </DataContext.StateProvider>
  );
}

App.propTypes = {
  testMode: PropTypes.bool
};
