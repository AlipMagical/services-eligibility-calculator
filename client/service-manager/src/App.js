/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import './i18n';

import { BrowserRouter, Redirect, Route, Switch } from 'react-router-dom';
import React, { createContext, useState } from 'react';
import {
  getLoggedInUser,
  getMillisUntilExpiration,
  isLoggedIn
} from './common/auth';

import { ANALYTICS_ID } from './common/config';
import { BASEPATH } from './common/config';
import Header from './components/Header';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import Toast from './components/Toast';
import departmentRoutes from './routes/department';
import getUserConfirmation from './getUserConfirmation';
import interestRoutes from './routes/interest';
import serviceRoutes from './routes/service';
import settingsRoutes from './routes/settings';
import { useSettings } from './hooks';

export const AuthContext = createContext();
export const ToastContext = createContext();
export const SettingsContext = createContext();

export default function App(props) {
  ReactGA.initialize(ANALYTICS_ID, { testMode: props.testMode });

  const [loginState, setLoginState] = useState({
    isLoggedIn: isLoggedIn(),
    loggedInUser: getLoggedInUser()
  });

  const onLoginStatusChanged = () => {
    //setTimeout to update when token might expire.
    const millis = getMillisUntilExpiration();
    millis && setTimeout(onLoginStatusChanged, millis);

    setLoginState({
      isLoggedIn: isLoggedIn(),
      loggedInUser: getLoggedInUser()
    });
  };

  const { settings, setSettings } = useSettings();

  const [toastMessage, setToastMessage] = useState(null);

  return (
    <AuthContext.Provider value={{ ...loginState, onLoginStatusChanged }}>
      <SettingsContext.Provider value={{ settings, setSettings }}>
        <ToastContext.Provider value={{ toastMessage, setToastMessage }}>
          <BrowserRouter getUserConfirmation={getUserConfirmation}>
            <Header />
            <div className="container bd-content">
              <Switch>
                {serviceRoutes}
                {settingsRoutes}
                {interestRoutes}
                {departmentRoutes}
                <Redirect from={BASEPATH} exact to={BASEPATH + 'services/'} />
                <Route render={() => <h1>Not Found</h1>} />
              </Switch>
            </div>
            {toastMessage && <Toast message={toastMessage} />}
          </BrowserRouter>
        </ToastContext.Provider>
      </SettingsContext.Provider>
    </AuthContext.Provider>
  );
}

App.propTypes = {
  testMode: PropTypes.bool
};
