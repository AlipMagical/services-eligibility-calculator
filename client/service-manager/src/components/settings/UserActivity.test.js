/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MemoryRouter, Route } from 'react-router-dom';
import { axe, toHaveNoViolations } from 'jest-axe';

import { AuthContext } from '../../App';
import React from 'react';
import { UserActivity } from '.';
import { act } from 'react-dom/test-utils';
import { mockUserActivitiesJson } from '../../../../testing/mockData';
import { render } from 'react-dom';

expect.extend(toHaveNoViolations);

it('renders user activity settings view', async () => {
  const container = document.createElement('div');

  fetch.resetMocks();
  fetch.doMockOnceIf(
    url => url.includes('/api/user_activities'),
    mockUserActivitiesJson
  );

  await act(async () => {
    return render(
      <MemoryRouter>
        <AuthContext.Provider
          value={{
            isLoggedIn: true,
            loggedInUser: { username: 'admin', roles: ['ROLE_ADMIN'] }
          }}
        >
          <Route component={props => <UserActivity {...props} />} />
        </AuthContext.Provider>
      </MemoryRouter>,
      container
    );
  });

  expect(container).toMatchSnapshot();
  expect(await axe(container)).toHaveNoViolations();
});
