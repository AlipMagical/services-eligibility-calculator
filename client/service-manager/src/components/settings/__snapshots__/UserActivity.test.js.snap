// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders user activity settings view 1`] = `
<div>
  <nav
    class="service-header navbar navbar-expand navbar-light sticky-top"
  >
    <div
      class="w-100"
    >
      <h1
        class="float-left"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root material-icons arrow-back"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            arrow_back
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        Settings
      </h1>
    </div>
  </nav>
  <div
    class="settings-tabs d-flex"
  >
    <span
      class="tab col-2"
    >
      <a
        href="/settings/manage"
      >
        Service management
      </a>
      <div
        class="tab-border"
      />
    </span>
    <span
      class="tab col-2"
    >
      <a
        href="/settings/service-activity"
      >
        Service activity
      </a>
      <div
        class="tab-border"
      />
    </span>
    <span
      class="tab col-2 active"
    >
      <a
        href="/settings/user-activity"
      >
        User activity
      </a>
      <div
        class="tab-border"
      />
    </span>
  </div>
  <table
    class="table table-fixed"
  >
    <thead>
      <tr>
        <th
          style="cursor: pointer; width: 40%;"
        >
          name
          <i
            class="material-icons"
          >
            arrow_downward
          </i>
        </th>
        <th
          style="cursor: pointer; width: 12%;"
        >
          logins
        </th>
        <th
          style="cursor: pointer; width: 12%;"
        >
          last logged in
        </th>
        <th
          style="cursor: pointer; width: 12%;"
        >
          edited
        </th>
        <th
          style="cursor: pointer; width: 12%;"
        >
          published
        </th>
        <th
          style="cursor: pointer; width: 12%;"
        >
          archived
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          admin
        </td>
        <td>
          2
        </td>
        <td>
          Feb 27, 2020
        </td>
        <td>
          0
        </td>
        <td>
          0
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          guest1
        </td>
        <td>
          0
        </td>
        <td />
        <td>
          1
        </td>
        <td>
          0
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          user
        </td>
        <td>
          6
        </td>
        <td>
          Feb 27, 2020
        </td>
        <td>
          1
        </td>
        <td>
          1
        </td>
        <td>
          0
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
